pilot/pkg/model/mesh.go

```diff
@@ -0,0 +1,785 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package model provides Mesh which is a platform independent
+// abstraction used by Controllers to synchronize the list of endpoints
+// used by the pilot with those available in the platform specific
+// registry. It implements all the necessary logic that's used for service
+// discovery based on routing rules and endpoint subsets.
+// Typical Usage:
+//
+//   import "istio.io/istio/pilot/model"
+//
+//   type MyPlatformController struct {
+//     Mesh
+//   }
+//   ...
+//   pc := MyPlatformController{model.NewMesh()}
+//   ...
+//   var allEndpoints []*model.ServiceInstances
+//   nativeEndpoints = buildYourPlatformServiceEndpointList()
+//   meshEndpoints := make([]*Endpoint, len(nativeEndpoints))
+//	 for idx, nativeEp := range nativeEndpoints {
+//     // Create mesh endpoint from relevant values of nativeEp
+//	   meshEndpoints[idx] = NewEndpoint(......)
+//   }
+//   pc.Reconcile(meshEndpoints)
+//
+package model
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"net"
+	"reflect"
+	"regexp"
+	"sort"
+	"strconv"
+	"sync"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/api"
+	types "github.com/gogo/protobuf/types"
+	multierror "github.com/hashicorp/go-multierror"
+	route "istio.io/api/routing/v1alpha2"
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	// The internal filter name for attributes that should be used for
+	// constructing Istio attributes on the Endpoint in
+	// <a href='https://goo.gl/mFzjpp'>reverse DNS form</a>.
+	istioConfigFilter = "io.istio.istio-config"
+
+	// Key Istio attribute names for mapping endpoints to subsets.
+	// For details, please see
+	// https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html
+
+	// UID is the attribute name for the mesh unique, platform-specific
+	// unique identifier for the server instance of the destination service. Mesh
+	// uses the label to uqniquely identify the endpoint when it receives updates
+	// from the Controller. Example: kubernetes://my-svc-234443-5sffe.my-namespace
```

Comment:
This does not look like a UID, only cluster-scoped ID.

Reply:
Nope - "destination-uid" has a very specific meaning that cannot be fudged :-) Let's please stick to https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html